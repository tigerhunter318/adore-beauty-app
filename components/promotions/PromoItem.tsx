import React, { useCallback, useState } from 'react'
import { StyleSheet } from 'react-native'
import { sanitizeContent } from '../../utils/format'
import { px, vw } from '../../utils/dimensions'
import Container from '../ui/Container'
import Type from '../ui/Type'
import ResponsiveImage from '../ui/ResponsiveImage'
import CustomCarousel from '../ui/CustomCarousel'
import RichTextContentDescription from '../RichText/RichTextContentDescription'
import useUrlNavigation from '../../navigation/utils/useUrlNavigation'
import theme from '../../constants/theme'
import Clipboard from '../../services/clipboard'
import Toast from '../../services/toast'
import { isValidObject } from '../../utils/validation'
import { getIn } from '../../utils/getIn'
import { formatPromotionData, formatPromotionDescription } from './utils/helpers'
import ImageSize from '../../constants/ImageSize'

const { width: productImageWidth, height: productImageHeight } = ImageSize.product.medium

const styles = StyleSheet.create({
  promoContainer: {
    borderWidth: 0.5,
    borderStyle: 'dashed',
    borderRadius: 1
  },
  image: {
    resizeMode: 'contain',
    width: 160,
    height: 65
  },
  productImage: {
    resizeMode: 'contain',
    width: 250,
    height: 250
  },
  imageContainer: {
    alignItems: 'center'
  },
  shopButtonContainer: {
    backgroundColor: theme.black,
    paddingHorizontal: 20,
    paddingVertical: 10
  },
  shopButtonText: {
    fontSize: 16,
    color: theme.white,
    textTransform: 'uppercase'
  },
  code: {
    paddingHorizontal: 10,
    backgroundColor: theme.codeOrange,
    lineHeight: 28,
    letterSpacing: 3,
    fontSize: 15,
    textAlign: 'center'
  },
  title: {
    textTransform: 'uppercase',
    textAlign: 'center',
    letterSpacing: 1.5,
    lineHeight: 20
  }
})

const getBtnLabel = (redeemUrl: string, hasCouponCode: any) => {
  if (redeemUrl && redeemUrl !== '/') {
    if (hasCouponCode) {
      return 'COPY CODE & SHOP NOW'
    }
    return 'SHOP NOW'
  }

  if (hasCouponCode) {
    return 'COPY CODE'
  }

  return ''
}

const logoStyles = {
  image: { ...styles.image },
  container: { ...styles.imageContainer }
}

const imageStyles = {
  image: { ...styles.productImage },
  container: { ...styles.imageContainer }
}

const PromoCopyButton = ({ onPress, isLoading, label }: { onPress: () => void; isLoading: boolean; label: string }) => (
  <Container center mt={2}>
    <Container style={styles.shopButtonContainer} onPress={onPress} disabled={isLoading} loading={isLoading}>
      <Type semiBold style={styles.shopButtonText}>
        {label}
      </Type>
    </Container>
  </Container>
)

type SimplePromoProps = {
  data: {} | any
  hasCouponCode: boolean
  onPressPromoItem: () => void
  isLoading: boolean
}

const SimplePromo = ({ data, hasCouponCode, onPressPromoItem, isLoading }: SimplePromoProps) => {
  const { code, brandLogo, message, siteMessage, image, redeemUrl, amount, title } = data
  const btnLabel = getBtnLabel(redeemUrl, hasCouponCode)
  const text = siteMessage || message

  return (
    <Container pv={3.35} ph={2}>
      {!!brandLogo && <ResponsiveImage src={brandLogo} styles={logoStyles} width={230} height={65} useAspectRatio />}
      {!!title && (
        <Type bold style={styles.title} mt={3}>
          {title}
        </Type>
      )}
      {!!text && (
        <Container mt={1} center>
          <RichTextContentDescription content={sanitizeContent(text)} align="left" />
        </Container>
      )}
      {!!image && (
        <Container mt={2}>
          <ResponsiveImage
            src={image}
            styles={imageStyles}
            width={productImageWidth}
            height={productImageHeight}
            useAspectRatio
          />
        </Container>
      )}
      {hasCouponCode && (
        <Type heading center letterSpacing={1} mt={1}>
          USE CODE: <Type color={theme.orange}>{code}</Type>
        </Type>
      )}
      {!!btnLabel && (
        <PromoCopyButton label={amount ? btnLabel : 'Sold out'} isLoading={isLoading} onPress={onPressPromoItem} />
      )}
    </Container>
  )
}

type ComplexPromoCarouselItemProps = {
  item: {}
  btnLabel: string
  amount: number | string
  isLoading: boolean
  onPressPromoItem: () => void
  style?: any
}

const ComplexPromoCarouselItem = ({
  item,
  btnLabel,
  amount,
  isLoading,
  onPressPromoItem,
  style = {}
}: ComplexPromoCarouselItemProps) => {
  const title = getIn(item, 'name')
  const description = formatPromotionDescription(getIn(item, 'description'))
  const brandLogo = getIn(item, 'brands.0.brand.image_link')
  const image = getIn(item, 'images.0.image.url_relative')

  return (
    <Container mt={2} style={style}>
      {!!image && (
        <ResponsiveImage
          src={image}
          styles={imageStyles}
          width={productImageWidth}
          height={productImageHeight}
          useAspectRatio
        />
      )}
      {!!brandLogo && <ResponsiveImage src={brandLogo} styles={logoStyles} width={230} height={65} useAspectRatio />}
      {!!title && (
        <Type bold style={styles.title} mt={3}>
          {title}
        </Type>
      )}
      {!!description && (
        <Container mt={2} center>
          <RichTextContentDescription content={sanitizeContent(description)} align="left" />
        </Container>
      )}
      {!!btnLabel && (
        <PromoCopyButton label={amount ? btnLabel : 'Sold out'} isLoading={isLoading} onPress={onPressPromoItem} />
      )}
    </Container>
  )
}

type ComplexPromoProps = {
  data: {} | any
  hasCouponCode: any
  onPressPromoItem: any
  isLoading: any
}

const ComplexPromo = ({ data, hasCouponCode, onPressPromoItem, isLoading }: ComplexPromoProps) => {
  const { code, message, siteMessage, title, products, redeemUrl, amount } = data
  const btnLabel = getBtnLabel(redeemUrl, hasCouponCode)
  const text = siteMessage || message

  const renderItem = useCallback(
    ({ item }) => (
      <ComplexPromoCarouselItem
        item={item}
        btnLabel={btnLabel}
        amount={amount}
        onPressPromoItem={onPressPromoItem}
        isLoading={isLoading}
        style={{ paddingHorizontal: 10 }}
      />
    ),
    []
  )

  return (
    <Container pv={3.35}>
      <Container ph={2}>
        {!!title && (
          <Type heading bold center size={24} letterSpacing={8} lineHeight={26}>
            {title}
          </Type>
        )}
        {!!text && (
          <Container mt={1.5} center>
            <RichTextContentDescription content={sanitizeContent(text)} align="left" />
          </Container>
        )}
        {hasCouponCode && (
          <Container mt={1.1} center>
            <Type bold style={styles.code}>
              use code{' '}
              <Type size={15} bold color={theme.orange}>
                {code}
              </Type>
            </Type>
          </Container>
        )}
      </Container>
      <Container ph={1}>
        <CustomCarousel items={products} containerHeight={px(600)} renderItem={renderItem} sliderWidth={vw(100) - 52} />
      </Container>
    </Container>
  )
}

const PromoItem = ({ item }: any) => {
  const urlNavigation = useUrlNavigation()
  const [loadingLinkUrl, setLoadingLinkUrl] = useState(null)
  const data = formatPromotionData(item)
  const { code, isAddFreeItem, products, redeemUrl } = data
  const hasCouponCode = !isAddFreeItem && code
  const promoType = products?.length > 2 ? 2 : 1

  const handlePressPromoItem = async () => {
    if (hasCouponCode) {
      Toast.show('Code copied', {
        duration: 1500,
        onShow: () => {
          Clipboard.copyToClipboard(code)
        }
      })
    }

    if (redeemUrl && redeemUrl !== '/' && redeemUrl !== loadingLinkUrl) {
      setLoadingLinkUrl(redeemUrl)
      await urlNavigation.push(redeemUrl)
      setLoadingLinkUrl(null)
    }
  }

  if (!isValidObject(data)) return null

  return (
    <Container style={styles.promoContainer}>
      {promoType === 1 ? (
        <SimplePromo
          data={data}
          hasCouponCode={hasCouponCode}
          onPressPromoItem={handlePressPromoItem}
          isLoading={redeemUrl && redeemUrl === loadingLinkUrl}
        />
      ) : (
        <ComplexPromo
          data={data}
          hasCouponCode={hasCouponCode}
          onPressPromoItem={handlePressPromoItem}
          isLoading={redeemUrl && redeemUrl === loadingLinkUrl}
        />
      )}
    </Container>
  )
}

export default PromoItem
