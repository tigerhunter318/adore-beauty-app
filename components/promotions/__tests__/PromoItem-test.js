import React from 'react'
import renderer from 'react-test-renderer'
import PromoItem from '../PromoItem'

const mockData = require('./__mocks__/PromoItem.json')

it(`PromoItem renders content`, () => {
  const tree = renderer.create(<PromoItem item={mockData} />).toJSON()

  expect(tree).toMatchSnapshot()
})
