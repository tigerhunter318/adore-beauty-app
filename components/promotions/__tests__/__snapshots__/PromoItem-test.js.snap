// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromoItem renders content 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "borderRadius": 1,
        "borderStyle": "dashed",
        "borderWidth": 0.5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingBottom": 33.5,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 33.5,
        },
        Object {},
      ]
    }
  >
    <View
      style={
        Array [
          Object {},
          Object {
            "alignItems": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "height": "auto",
                "width": "100%",
              },
              Object {
                "aspectRatio": 3.5384615384615383,
                "height": 65,
                "resizeMode": "contain",
                "width": 160,
              },
            ],
          ]
        }
      >
        <FastImageView
          defaultSource={null}
          imageProps={Object {}}
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.adorebeauty.com.au/pim_brand_logo/000/002/156/Jo_Loves_Logo_160_x_65.png?w=460&h=130&fmt=webp",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
    <Text
      allowFontScaling={false}
      mt={3}
      style={
        Array [
          Object {
            "fontFamily": "Montserrat-Bold",
            "marginTop": 30,
          },
          Object {
            "letterSpacing": 1.5,
            "lineHeight": 20,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
        ]
      }
    >
      Jo Loves Jo by Jo Loves 70g Candle*
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginTop": 10,
          },
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <View>
          <Text
            onLongPress={null}
            onPress={null}
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "fontFamily": "Montserrat-Regular",
                  },
                  Object {
                    "color": "#404040",
                    "fontSize": 13,
                    "letterSpacing": 0.1,
                    "lineHeight": 24,
                    "marginBottom": 5,
                    "paddingHorizontal": 0,
                    "textAlign": "left",
                  },
                ]
              }
            >
              Your gift when you spend $90 or more on Jo Loves* While Stocks Last
            </Text>
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginTop": 20,
          },
          Object {},
        ]
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#000000",
            "opacity": 1,
            "paddingHorizontal": 20,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "fontFamily": "Montserrat-SemiBold",
              },
              Object {
                "color": "#FFFFFF",
                "fontSize": 16,
                "textTransform": "uppercase",
              },
            ]
          }
        >
          SHOP NOW
        </Text>
      </View>
    </View>
  </View>
</View>
`;
